components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs

```diff
@@ -760,43 +773,52 @@ internal sealed class SubscriptionWrapper : ISubscription {
-                                        heartbeatValue.PublishTime = publishTime;
-                                        if (message.Notifications == null) {
-                                            message.Notifications =
-                                                new List<MonitoredItemNotificationModel>();
-                                        }
-                                        message.Notifications.Add(heartbeatValue);
-                                    }
-                                    continue;
+                                Exists(m => m.ClientHandle == item.Item.ClientHandle)
+                                && item.ValidateHeartbeat(publishTime)) {
+
+                                MonitoredItemNotificationModel GetDefaultNotification()
+                                {
+                                    return new MonitoredItemNotificationModel {
+                                        Id = item?.Template?.Id,
+                                        DisplayName = item.Item.DisplayName,
+                                        NodeId = item?.Template?.StartNodeId,
+                                        AttributeId = item.Item.AttributeId,
+                                        ClientHandle = item.Item.ClientHandle,
+                                        Value = new DataValue(Variant.Null,
+                                            item.Item?.Status?.Error?.StatusCode ??
+                                            StatusCodes.BadMonitoredItemIdInvalid),
+                                        Overflow = false,
+                                        NotificationData = null,
+                                        StringTable = null,
+                                        DiagnosticInfo = null,
+                                    };
                                 }
+
+                                var heartbeatValue = item.Item?.LastValue.
+                                    ToMonitoredItemNotification(item.Item, GetDefaultNotification);
+                                if (heartbeatValue != null) {
+                                    heartbeatValue.SequenceNumber = sequenceNumber;
+                                    heartbeatValue.IsHeartbeat = true;
+                                    heartbeatValue.PublishTime = publishTime;
+                                    message.Notifications?.Add(heartbeatValue);
+                                }
+                                continue;
                             }
                             item.ValidateHeartbeat(publishTime);
                         }
                     }
 
+                    var erroneousNotifications = message.Notifications?
+                        .Where(n => n.Value.Value == null
+                            || StatusCode.IsNotGood(n.Value.StatusCode))
+                        .ToList();
+
+                    if (erroneousNotifications.Any()) {
```

Comment:
Since `message.Notifications?` may be null, we need to check for null here as well by `erroneousNotifications?.Any()`.

Reply:
OK, I take this back as we generate `message` and make sure that `Notifications` is not null during construction. So we can even take `?` out from `message.Notifications?` when initializing `erroneousNotifications`. But it is fine as it is as well.